ui/pages/confirmations/components/confirm/row/dataTree.tsx

```diff
@@ -20,49 +26,127 @@ export type TreeData = {
   type: string;
 };
 
+const FIELD = {
+  AMOUNT: 'amount',
+  BUY_AMOUNT: 'buyAmount',
+  DEADLINE: 'deadline',
+  END_AMOUNT: 'endAmount',
+  END_TIME: 'endTime',
+  EXPIRATION: 'expiration',
+  SELL_AMOUNT: 'sellAmount',
+  SIG_DEADLINE: 'sigDeadline',
+  START_AMOUNT: 'startAmount',
+  START_TIME: 'startTime',
+  VALID_TO: 'validTo',
+  VALUE: 'value',
+};
+
+const FIELD_TOKEN_UTILS_PRIMARY_TYPES = {
+  [FIELD.AMOUNT]: [...PRIMARY_TYPES],
+  [FIELD.BUY_AMOUNT]: [PRIMARY_TYPE.ORDER, PRIMARY_TYPE.ORDER_COMPONENTS],
+  [FIELD.END_AMOUNT]: [PRIMARY_TYPE.ORDER, PRIMARY_TYPE.ORDER_COMPONENTS],
+  [FIELD.SELL_AMOUNT]: [PRIMARY_TYPE.ORDER, PRIMARY_TYPE.ORDER_COMPONENTS],
+  [FIELD.START_AMOUNT]: [PRIMARY_TYPE.ORDER, PRIMARY_TYPE.ORDER_COMPONENTS],
+  [FIELD.VALUE]: [...PRIMARY_TYPES],
+};
+
+const FIELD_DATE_PRIMARY_TYPES = {
+  [FIELD.DEADLINE]: [...PRIMARY_TYPES],
+  [FIELD.END_TIME]: [PRIMARY_TYPE.ORDER, PRIMARY_TYPE.ORDER_COMPONENTS],
+  [FIELD.EXPIRATION]: [PRIMARY_TYPE.PERMIT_BATCH, PRIMARY_TYPE.PERMIT_SINGLE],
+  [FIELD.SIG_DEADLINE]: [
+    PRIMARY_TYPE.PERMIT_BATCH,
+    PRIMARY_TYPE.PERMIT_BATCH_TRANSFER_FROM,
+    PRIMARY_TYPE.PERMIT_SINGLE,
+    PRIMARY_TYPE.PERMIT_TRANSFER_FROM,
+  ],
+  [FIELD.START_TIME]: [PRIMARY_TYPE.ORDER, PRIMARY_TYPE.ORDER_COMPONENTS],
+  [FIELD.VALID_TO]: [PRIMARY_TYPE.ORDER, PRIMARY_TYPE.ORDER_COMPONENTS],
+};
+
+const getTokenDecimalsOfDataTree = async (
+  dataTreeData: Record<string, TreeData> | TreeData[],
+): Promise<void | number> => {
+  if (Array.isArray(dataTreeData)) {
+    return undefined;
+  }
+
+  const tokenContract = (dataTreeData as Record<string, TreeData>).token
+    ?.value as string;
+  if (!tokenContract) {
+    return undefined;
+  }
+
+  const decimals = parseInt(
+    (await getTokenStandardAndDetails(tokenContract)).decimals ?? '0',
+    10,
+  );
+
+  return decimals;
+};
```

Comment:
May be useful to extract above constants and utility into separate file.

Reply:
unsure how useful this is yet since the only use-case lives here. Will leave it here for now. Let me know if there are other reasons I may be missing.